bitcoin/src/blockdata/witness.rs

```diff
@@ -319,6 +320,26 @@ impl Witness {
         self.content[end_varint..end_varint + new_element.len()].copy_from_slice(new_element);
     }
 
+    /// Pushes data onto the witness stack as required to spend a P2WPKH output.
+    ///
+    /// Pushes a DER-encoded ECDSA signature with a signature hash type as a new element on the
+    /// witness, followed by the serialized public key.
+    ///
+    /// Also useful for spending a P2SH-P2WPKH output.
+    pub fn push_p2wpkh(
+        &mut self,
+        signature: &ecdsa::SerializedSignature,
+        hash_type: EcdsaSighashType,
+        pubkey: &PublicKey,
+    ) {
+        self.push_bitcoin_signature(signature, hash_type);
+        if pubkey.compressed {
+            self.push_slice(&pubkey.inner.serialize())
+        } else {
+            self.push_slice(&pubkey.inner.serialize_uncompressed())
+        }
```

Comment:
Segwit v0 forbids uncompressed keys for use in signature operations.

This is a big enough foot gun that I think we should somehow prevent it.

Reply:
Oh boy, my bad, thanks man. I'll fix it up.

Reply:
I whipped this up in case we want to allow the footgun but at least warn the caller.

This could probably be brushed up and made prettier with a nice error variant name containing the closure with a name that makes it clear why it's a footgun.

But tbh it's probably just better to fail on uncompressed.

```rust
    pub fn push_p2wpkh<'a>(
        &'a mut self,
        signature: &'a ecdsa::SerializedSignature,
        hash_type: EcdsaSighashType,
        pubkey: &'a PublicKey,
    ) -> Result<(), impl FnMut() + 'a> {
        let mut closure = move || {
            self.push_bitcoin_signature(signature, hash_type);
            if pubkey.compressed {
                self.push_slice(&pubkey.inner.serialize())
            } else {
                self.push_slice(&pubkey.inner.serialize_uncompressed())
            }
        };
        if pubkey.compressed {
            closure();
            Ok(())
        } else {
            Err(closure)
        }
    }
```

Reply:
Or didn't we add a `CompressedPublicKey` type a while back (I forget)

Reply:
I added a `UncompressPubkeyError` type, and the following check
```rust
        if !pubkey.compressed {
            return Err(UncompressedPubkeyError);
        }
```
I would have thought this sort of check would exist in other places but my quick greping couldn't find one. The error type is currently defined in `witness` but might be better in `crypto::key`?

Reply:
We are trying to "hide" the legacy key stuff but so far haven't come up with an elegant solution.